File: OpenMP/solutions/omp_pi_critical.c
#include <omp.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

int main (int argc, char *argv[])
{

  //initialize variables
  int i;
  double pi = 0;
  int niter = 100000000;

  // Get timing
  double start,end;
  start=omp_get_wtime();

  // Calculate PI using Leibnitz sum
  /* Fork a team of threads */
#pragma omp parallel for
  for(i = 0; i < niter; i++)
  {
#pragma omp critical
{
    pi = pi + pow(-1, i) * (4 / (2*((double) i)+1));
}
  }

  // Stop timing
  end=omp_get_wtime();

  // Print result
  printf("Pi estimate: %.20f, obtained in %f seconds\n", pi, end-start);
}

File: OpenMP/solutions/omp_pi_critical.f
      program pi

        use omp_lib
        implicit none

c       initialize variables
        integer :: i
        double precision :: pi_val = 0.d0
        integer :: niter = 100000000

c       Get timing
        double precision :: start,end
        start=omp_get_wtime()

c       Calculate PI using Leibnitz sum
!$OMP PARALLEL DO
        do i=0,niter-1
!$OMP CRITICAL
          pi_val = pi_val + (-1)**i * (4 / (2*dble(i)+1))
!$OMP END CRITICAL
        enddo
!$OMP END PARALLEL DO

c       Stop timing
        end=omp_get_wtime()-start;

c       Print result
        print *,"Pi estimate: ",pi_val," obtained in ",end," seconds"

      end program
